File: python/cards.py
"""
Representations for cards and decks.
"""
from enum import Enum
import itertools
import random

SUITS = {
    1: ['♠', 'SPADES'],
    2: ['♥', 'HEARTS'],
    3: ['♦', 'DIAMONDS'],
    4: ['♣', 'CLUBS'],
}

Suit = Enum(
    value='Suit',
    names=itertools.chain.from_iterable(
        itertools.product(v, [k]) for k, v in SUITS.items()
    )
)

def random_suit():
    """
    Returns a random suit.
    """
    return random.choice(list(Suit))

class Card:
    """
    Represents a card. Requires a suit and a rank.
    """
    def __init__(self, suit, rank):
        if not suit in set(x for x in Suit):
            raise RuntimeError('Illegal suit ("' + str(suit) + '")')
        if rank not in range(1, 14):
            raise RuntimeError('Illegal rank ("' + str(rank) + '")')
        self.suit = suit
        self.rank = rank
        self.value = min(rank, 10)

    def __str__(self):
        if self.rank == 1:
            name = 'A'
        elif self.rank == 11:
            name = 'J'
        elif self.rank == 12:
            name = 'Q'
        elif self.rank == 13:
            name = 'K'
        else:
            name = str(self.rank)
        return name + self.suit.name

    def __repr__(self):
        return str(self)

    def __lt__(self, other):
        return self.rank < other.rank

class Deck:
    """
    Represents a deck of cards.
    """
    def __init__(self):
        self.cards = [Card(suit, rank) for rank in range(1, 14) for suit in Suit]

    def __str__(self):
        return 'Deck: ' + str(self.cards)

    def __len__(self):
        return len(self.cards)

    def shuffle(self):
        """
        Randomizes the order of the deck.
        """
        random.shuffle(self.cards)

    def draw(self):
        """
        Removes and returns the top card of the deck.
        """
        return self.cards.pop()

File: src/server/game/cards.ts
/**
 * @brief Representations for cards and decks.
 */
import { Util } from './util';

/**
 * @brief Represents a card's suit.
 *
 * The Suit.SUITS array contains all the possible values for suits.
 */
export class Suit {
    public static readonly SPADES = new Suit('♠', 'SPADES');

    public static readonly HEARTS = new Suit('♥', 'HEARTS');

    public static readonly DIAMONDS = new Suit('♦', 'DIAMONDS');

    public static readonly CLUBS = new Suit('♣', 'CLUBS');

    public static readonly SUITS = Object.freeze([
        Suit.SPADES,
        Suit.HEARTS,
        Suit.DIAMONDS,
        Suit.CLUBS,
    ]);

    public static readonly randomSuit = () => {
        return Util.arraySample(Suit.SUITS);
    };

    private constructor(
        public readonly icon: string,
        public readonly name: string,
    ) {}

    public readonly toString = () => {
        return `${this.icon} (${this.name})`;
    };

    public readonly equals = (suit: Suit) => {
        return this.icon === suit.icon && this.name === suit.name;
    };
}

/**
 * @brief All possible ranks for cards.
 */
type Rank = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13;

/**
 * @brief Ranks enum.
 *
 * These are the actual runtime values for all the possible ranks.
 */
const RANKS: Rank[] = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13];

/**
 * @brief Represents a card. Requires a suit and a rank.
 */
export class Card {
    public readonly value: number;

    public constructor(public readonly suit: Suit, public readonly rank: Rank) {
        if (
            !Suit.SUITS.some((s) => {
                return s.equals(suit);
            })
        ) {
            throw new Error(`Illegal suit (${suit.icon}, ${suit.name})`);
        }
        if (!(rank >= 1 && rank <= 13)) {
            throw new Error(`Illegal rank ("${rank}")`);
        }
        this.value = Math.min(rank, 10);
    }

    public readonly toString = () => {
        return `${(() => {
            switch (this.rank) {
                case 1:
                    return 'A';
                case 11:
                    return 'J';
                case 12:
                    return 'Q';
                case 13:
                    return 'K';
                default:
                    return `${this.rank}`;
            }
        })()}${this.suit.icon}`;
    };
}

/**
 * @brief Represents a deck of cards.
 */
export class Deck {
    private cards = Suit.SUITS.flatMap((suit) => {
        return RANKS.map((rank) => {
            return new Card(suit, rank);
        });
    });

    public readonly isEmpty = () => {
        return this.cards.length <= 0;
    };

    public readonly draw = () => {
        if (this.isEmpty()) {
            throw new Error('Cannot draw from an empty deck');
        } else {
            return this.cards.pop() as Card;
        }
    };

    public readonly shuffle = () => {
        this.cards = Util.arrayShuffle(this.cards);
    };

    public readonly toString = () => {
        return `Deck: ${this.cards.map((card) => {
            return card.toString();
        })}`;
    };
}
